registry_test.go

```diff
@@ -0,0 +1,384 @@
+		// dimensions, users can't sometimes use a vector and sometimes use a la
+		// carte scalars.
+
+		// dims: foo, baz
+		_, err := r.NewCounterVector(Opts{
+			Name:           "ownership",
+			Help:           "help",
+			Labels:         Labels{"foo": "bar"},
+			VariableLabels: []string{"baz"},
+		})
+		require.NoError(t, err)
+
+		// same dims
+		_, err = r.NewCounter(Opts{
+			Name:   "ownership",
+			Help:   "help",
+			Labels: Labels{"foo": "bar", "baz": "quux"},
+		})
+		require.Error(t, err)
+	})
+
+	t.Run("duplicate scrubbed name", func(t *testing.T) {
+		// Uniqueness is enforced after the metric name is scrubbed.
+		_, err := r.NewCounterVector(Opts{
+			Name:           "scrubbed_name",
+			Help:           "help",
+			VariableLabels: []string{"bar"},
+		})
+		assert.NoError(t, err, "Failed to register new metric.")
+		_, err = r.NewCounterVector(Opts{
+			Name:           "scrubbed&name",
+			Help:           "help",
+			VariableLabels: []string{"bar"},
+		})
+		assert.Error(t, err)
+	})
+
+	t.Run("duplicate scrubbed dimensions", func(t *testing.T) {
+		// Uniqueness is enforced after labels are scrubbed.
+		_, err := r.NewCounterVector(Opts{
+			Name:           "scrubbed_dimensions",
+			Help:           "help",
+			Labels:         Labels{"b_r": "baz"},
+			VariableLabels: []string{"q__x"},
+		})
+		assert.NoError(t, err, "Failed to register new metric.")
+		_, err = r.NewCounterVector(Opts{
+			Name:           "scrubbed_dimensions",
+			Help:           "help",
+			Labels:         Labels{"b&r": "baz"},
+			VariableLabels: []string{"q&&x"},
+		})
+		assert.Error(t, err)
+	})
+
+	t.Run("constant label name specified twice", func(t *testing.T) {
+		// Within a single user-supplied set of constant labels, scrubbing may not
+		// introduce duplicates.
+		_, err = r.NewCounterVector(Opts{
+			Name:           "user_error_constant_labels",
+			Help:           "help",
+			Labels:         Labels{"b_r": "baz", "b&r": "baz"},
+			VariableLabels: []string{"quux"},
+		})
+		assert.Error(t, err)
+	})
+
+	t.Run("variable label name specified twice", func(t *testing.T) {
+		// Within a single user-supplied set of variable labels, scrubbing may not
+		// introduce duplicates.
+		_, err = r.NewCounterVector(Opts{
+			Name:           "user_error_variable_labels",
+			Help:           "help",
+			VariableLabels: []string{"f__", "f&&"},
+		})
+		assert.Error(t, err)
+	})
+}
+
+func TestLabeledPrecedence(t *testing.T) {
+	r, c := New()
+	r = r.Labeled(Labels{"foo": "bar"}).Labeled(Labels{"foo": "baz"})
+	_, err := r.NewCounter(Opts{
+		Name: "test_counter",
+		Help: "help",
+	})
+	require.NoError(t, err, "Failed to create counter.")
+	snap := c.Snapshot()
+	require.Equal(t, 1, len(snap.Counters), "Unexpected number of counters.")
+	assert.Equal(t, SimpleSnapshot{
+		Name:   "test_counter",
+		Labels: Labels{"foo": "baz"},
+	}, snap.Counters[0], "Unexpected counter snapshot.")
+}
+
+func TestLabeledAutoScrubbing(t *testing.T) {
```

Comment:
Can we add another test for calling `Labeled` on a registry twice, and adding the same constLabel name twice?

Reply:
Sure.